### DISCUSSION TITLE
`hipmallocAsync` and hipSYCL

### DISCUSSION THREAD

**Discussion by csccva** [AdaptiveCpp/AdaptiveCpp#1078] [General]
Hello,

I discovered a few months ago that `hipmallocAsync`  is not supported by rocm version < 5.3. Since all the  USM memory allocations in sycl are asynchronous, how was this achieved in hipsycl with previous rocm versions? Is there another asynchronous allocations function? Just curious :)

**Comment by illuhad** [comment]
In SYCL, all USM memory management functions are synchronous, not asynchronous :)

While you pass a `queue` to e.g. `malloc_device(..., q)`, this does not correspond to an operation that is actually enqueued. Rather this should be only seen as a convenience shortcut for `malloc_device(..., q.get_device(), q.get_context())`, which also exists.
Only member functions of `queue` or `handler` actually correspond to asynchronous execution.

**Comment by csccva** [comment]
Ohh. I must have mixed the things. I must have seen the `.wait()` for some operations, but maybe it was the copy to device and not the malloc. So the program does not proceed to the next instructions until the allocation is done?

**Reply by illuhad** [reply]
Yes. It's the same as with the regular, non-async `cudaMalloc` or `hipMalloc`. An important consequence is also that you should always wait until the kernels are done before calling `sycl::free()`. Because `sycl::free()` is not enqueued and releases memory right away, you otherwise risk that memory is freed while it is in use by a kernel (which are executed asynchronously).

**Comment by csccva** [comment]
Thanks for the reply. This is quite important to keep in mind. I always assumed that the allocation and freeing are associated with a specific queue.
